File: dev_tools/gsheets2img/gsheets2img.py
import asyncio
import json
import shutil
import tempfile
import zipfile
from pathlib import Path

import aiohttp
from playwright.async_api import async_playwright

config_path = Path(__file__).parent / "config/default.json"
with open(config_path, "r") as f:
    config = json.load(f)

gsheets2img = config["gsheets2img"]
sheet_id = gsheets2img["sheetID"]
output_dir = Path(__file__).parent / Path(gsheets2img["outputDir"])
include_sheets = gsheets2img.get("includeSheets", [])
exclude_sheets = gsheets2img.get("excludeSheets", [])
concurrency = gsheets2img.get("concurrency", 2)

resolved_output_dir = output_dir.resolve()


async def download(sheet_id):
    temp_dir = Path(tempfile.mkdtemp(prefix="gs2imgz-"))
    zip_path = temp_dir / f"{sheet_id}.zip"

    url = f"https://docs.google.com/spreadsheets/d/{sheet_id}/export?format=zip"

    async with aiohttp.ClientSession() as session:
        async with session.get(url) as response:
            response.raise_for_status()
            with open(zip_path, "wb") as f:
                f.write(await response.read())

    return zip_path


async def unzip(zip_path):
    temp_dir = Path(tempfile.mkdtemp(prefix="gs2imgx-"))
    with zipfile.ZipFile(zip_path, "r") as zip_ref:
        zip_ref.extractall(temp_dir)
    shutil.rmtree(zip_path.parent, ignore_errors=True)
    return temp_dir


async def screenshot(html_path, png_path, browser):
    page = await browser.new_page(
        viewport={"width": 1920, "height": 1080}, device_scale_factor=2
    )
    await page.goto(f"file://{html_path}", timeout=0)

    row_header = await page.query_selector(".row-header-wrapper")
    row_header_box = await row_header.bounding_box()
    row_header_width = row_header_box["width"]

    tbody = await page.query_selector("tbody")
    bounding_box = await tbody.bounding_box()

    await page.set_viewport_size(
        {
            "width": int(bounding_box["width"]) + 10,
            "height": int(bounding_box["height"]) + 10,
        }
    )

    bounding_box["x"] += row_header_width + 1
    bounding_box["width"] -= row_header_width + 1

    await page.screenshot(path=str(png_path), clip=bounding_box)
    await page.close()


async def process_sheets(sheet_names, extracted_dir, browser):
    tasks = []
    semaphore = asyncio.Semaphore(concurrency)

    async def process_sheet(sheet_name):
        async with semaphore:
            file_name = sheet_name.replace(" ", "_") + ".jpeg"
            html_path = extracted_dir / f"{sheet_name}.html"
            png_path = resolved_output_dir / file_name
            print(f"Uploading {file_name}")
            await screenshot(html_path, png_path, browser)

    for sheet_name in sheet_names:
        tasks.append(process_sheet(sheet_name))

    await asyncio.gather(*tasks)


async def main():

    resolved_output_dir.mkdir(parents=True, exist_ok=True)

    zip_path = await download(sheet_id)
    extracted_dir = await unzip(zip_path)

    files = [f for f in extracted_dir.iterdir() if f.suffix == ".html"]
    sheet_names = [
        f.stem
        for f in files
        if (not include_sheets or f.stem in include_sheets)
        and f.stem not in exclude_sheets
    ]

    async with async_playwright() as playwright:
        browser = await playwright.firefox.launch()
        await process_sheets(sheet_names, extracted_dir, browser)
        await browser.close()

    shutil.rmtree(extracted_dir)


if __name__ == "__main__":
    asyncio.run(main())
    print("Finished Uploading Images!")

File: packages/dev/src/automated_imgur_upload/gsheets2img.ts
import { chromium } from "playwright"
import path from "path"
import fs from "fs/promises"
import JSZip from "jszip"
import os from "os"

const sheetId = "1wWMIzaUKISAXMbOEnmsuuLkO9PesabpdTUWdosvHygM"
const outputDir = "public/images/equip_misc/"
const includeSheets: string[] = []
const excludeSheets = [
  "(WiP) SS RLD Chart",
  "Copy of BB Guns",
  "(WiP) OpSi Image Guide Pt4",
  "Ammo Modifiers Chart",
  "(WiP) Rikka Specific Guide",
]

const download = async (sheetId: string): Promise<string> => {
  const tempDir = await fs.mkdtemp(path.join(os.tmpdir(), "gs2imgz-"))
  const zipPath = path.join(tempDir, `${sheetId}.zip`)
  const url = `https://docs.google.com/spreadsheets/d/${sheetId}/export?format=zip`
  const response = await fetch(url)
  if (!response.ok) {
    throw new Error(`Failed to download file: ${response.statusText}`)
  }
  const buffer = await response.arrayBuffer()
  await Bun.write(zipPath, buffer)
  return zipPath
}

const unzip = async (zipPath: string): Promise<string> => {
  const tempDir = await fs.mkdtemp(path.join(os.tmpdir(), "gs2imgx-"))
  const zipData = await Bun.file(zipPath).arrayBuffer()
  const zip = await JSZip.loadAsync(zipData)

  const extractionPromises: Promise<void>[] = []

  zip.forEach((relativePath, zipEntry) => {
    const outputPath = path.join(tempDir, relativePath)

    if (zipEntry.dir) {
      extractionPromises.push(fs.mkdir(outputPath, { recursive: true }))
    } else {
      const dir = path.dirname(outputPath)
      const promise = fs.mkdir(dir, { recursive: true }).then(() => {
        return zipEntry.async("nodebuffer").then((content) => {
          return fs.writeFile(outputPath, content)
        })
      })
      extractionPromises.push(promise)
    }
  })

  await Promise.all(extractionPromises)
  await fs.rm(path.dirname(zipPath), { recursive: true, force: true })
  return tempDir
}

const screenshot = async (htmlPath: string, pngPath: string, browser: any) => {
  const page = await browser.newPage({
    viewport: { width: 1920, height: 1080 },
    deviceScaleFactor: 2,
  })
  await page.goto(`file://${htmlPath}`, { timeout: 0 })

  const rowHeader = await page.querySelector(".row-header-wrapper")
  if (!rowHeader) {
    console.log(
      `Skipping ${path.basename(htmlPath)}: row-header-wrapper not found.`,
    )
    await page.close()
    return
  }

  const rowHeaderBox = await rowHeader.boundingBox()
  if (!rowHeaderBox) {
    console.log(`Skipping ${path.basename(htmlPath)}: bounding box not found.`)
    await page.close()
    return
  }

  const tbody = await page.querySelector("tbody")
  if (!tbody) {
    console.log(`Skipping ${path.basename(htmlPath)}: tbody not found.`)
    await page.close()
    return
  }

  const boundingBox = await tbody.boundingBox()
  if (!boundingBox) {
    console.log(`Skipping ${path.basename(htmlPath)}: bounding box not found.`)
    await page.close()
    return
  }

  const rowHeaderWidth = rowHeaderBox.width

  const clipArea = {
    x: boundingBox.x + rowHeaderWidth + 1,
    y: boundingBox.y,
    width: boundingBox.width - rowHeaderWidth - 1,
    height: boundingBox.height,
  }

  await page.setViewportSize({
    width: Math.max(1920, Math.round(clipArea.width) + 100),
    height: Math.max(1080, Math.round(clipArea.height) + 100),
  })

  console.log(`Uploading ${path.basename(htmlPath, ".html")}`)
  await page.screenshot({ path: pngPath, clip: clipArea })
  await page.close()
}

const processSheets = async (
  sheetNames: string[],
  extractedDir: string,
  browser: any,
) => {
  for (const sheetName of sheetNames) {
    const fileName = sheetName.replace(/ /g, "_") + ".jpeg"
    const htmlPath = path.join(extractedDir, `${sheetName}.html`)
    const pngPath = path.join(outputDir, fileName)
    if (!(await Bun.file(htmlPath).exists())) {
      console.log(`Skipping ${sheetName}: HTML file not found.`)
      continue
    }
    await screenshot(htmlPath, pngPath, browser)
  }
}

export const main = async () => {
  await fs.mkdir(outputDir, { recursive: true })
  const zipPath = await download(sheetId)
  const extractedDir = await unzip(zipPath)
  const files = await fs.readdir(extractedDir)
  const sheetNames = files
    .filter((f) => f.endsWith(".html"))
    .map((f) => path.basename(f, ".html"))
    .filter(
      (stem) =>
        (includeSheets.length === 0 || includeSheets.includes(stem)) &&
        !excludeSheets.includes(stem),
    )

  console.log(`Found ${sheetNames.length} sheets to process.`)

  const browser = await chromium.launch()
  try {
    await processSheets(sheetNames, extractedDir, browser)
  } finally {
    await browser.close()
  }

  await fs.rm(extractedDir, { recursive: true, force: true })
  console.log("Finished uploading images.")
}
